### ISSUE TITLE
Implication of layout equality for reference layout

### ISSUE THREAD

**Issue by joshlf** [rust-lang/unsafe-code-guidelines#375]
If I have two types, `T` and `U`, whose layouts are guaranteed to be the same (e.g., `#[repr(transparent)] struct T(U);`), is it guaranteed that the reference types `&T` and `&U` have the same layout (likewise for `&mut T` and `&mut U`)? Is it guaranteed that `transmute::<&T, &U>(...)` is sound?

What about for other reference-y types with well-defined layouts? E.g., `Box<T>`'s layout is well-defined for `T: Sized`.

**Comment by Lokathor** [comment]
The answer to all of these questions better be "yes" because people have been doing this in unsafe code for *quite* a while now.

**Comment by RalfJung** [comment]
> Is it guaranteed that transmute::<&T, &U>(...) is sound?

I assume you also mean the implicit lifetimes here to be the same (which `transmute` does not enforce), otherwise this is clearly unsound.

**Comment by RalfJung** [comment]
Also, the answer to "is that [sound](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#soundness-of-code--of-a-library)" is clearly wrong -- even if `T` and `U` have the same layout, they might have different user-defined invariants, so transmuting between them can cause UB down the road.

I assume you are asking whether the transmute itself is fine in terms of data layout; exposing such a transmute to safe code clearly requires further precautions.

**Comment by joshlf** [comment]
> I assume you are asking whether the transmute itself is fine in terms of data layout (i.e., it is defined behavior and data accesses behave in the expected way); exposing such a transmute to safe code clearly requires further precautions.

Yeah, that's my question. I'm not planning on exposing this as a generic operation to safe code or anything like that - rather, I want to know whether code like the following is sound (this is taken from [this zerocopy PR)](https://github.com/google/zerocopy/pull/122/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R2496-R2515):

```rust
#[repr(transparent)]
struct Unsized([u8]);

impl Unsized {
    fn from_mut_slice(slc: &mut [u8]) -> &mut Unsized {
        unsafe { core::mem::transmute(slc) }
    }
}
```
